openpype/hosts/maya/api/pipeline.py

```diff
@@ -681,6 +686,116 @@ def before_workfile_save(event):
+    else:
+        buttons = accept
+
+    state = QtWidgets.QMessageBox.warning(
+        None,
+        "",
+        message,
+        buttons=buttons,
+        defaultButton=accept
+    )
+
+    return state == accept
+
+
+def workfile_save_before_xgen(event):
+    """Manage Xgen external files when switching context.
+
+    Xgen has various external files that needs to be unique and relative to the
+    workfile, so we need to copy and potentially overwrite these files when
+    switching context.
+
+    Args:
+        event (Event) - openpype/lib/events.py
+    """
+    if not cmds.pluginInfo("xgenToolkit", query=True, loaded=True):
+        return
+
+    import xgenm
+
+    current_work_dir = legacy_io.Session["AVALON_WORKDIR"].replace("\\", "/")
+    expected_work_dir = event.data["workdir_path"].replace("\\", "/")
+    if current_work_dir == expected_work_dir:
+        return
+
+    palettes = cmds.ls(type="xgmPalette", long=True)
+    if not palettes:
+        return
+
+    transfers = []
+    overwrites = []
+    attribute_changes = {}
+    attrs = ["xgFileName", "xgBaseFile"]
+    for palette in palettes:
+        sanitized_palette = palette.replace("|", "")
+        project_path = xgenm.getAttr("xgProjectPath", sanitized_palette)
+        _, maya_extension = os.path.splitext(event.data["filename"])
+
+        for attr in attrs:
+            node_attr = "{}.{}".format(palette, attr)
+            attr_value = cmds.getAttr(node_attr)
+
+            if not attr_value:
+                continue
+
+            source = os.path.join(project_path, attr_value)
+
+            attr_value = event.data["filename"].replace(
+                maya_extension,
+                "__{}{}".format(
+                    sanitized_palette.replace(":", "__"),
+                    os.path.splitext(attr_value)[1]
+                )
+            )
+            target = os.path.join(expected_work_dir, attr_value)
+
+            transfers.append((source, target))
+            attribute_changes[node_attr] = attr_value
+
+        relative_path = xgenm.getAttr(
+            "xgDataPath", sanitized_palette
+        ).split(os.pathsep)[0]
+        absolute_path = relative_path.replace("${PROJECT}", project_path)
+        for root, _, files in os.walk(absolute_path):
+            for f in files:
+                source = os.path.join(root, f).replace("\\", "/")
+                target = source.replace(project_path, expected_work_dir + "/")
+                transfers.append((source, target))
+                if os.path.exists(target):
+                    overwrites.append(target)
+
+    # Ask user about overwriting files.
+    if overwrites:
+        msg = (
+            "WARNING! Potential loss of data.\n\n"
+            "Found duplicate Xgen files in new context.\n"
+            "Do you want to overwrite?\n\n{}".format("\n".join(overwrites))
+        )
+        accept = prompt_warning(msg, show_cancel=True)
+        if not accept:
+            return
```

Comment:
```suggestion
        if not lib.IS_HEADLESS:
            accept = prompt_warning(msg, show_cancel=True)
            if not accept:
                return
        else:
            print(msg)
```

Reply:
I'm up for overwriting by default, but just wanna make sure we consider other options.

If we raise an error, it would be up to the users to figure out why there are duplicate Xgen files and manage the files themselves.
If we built in some mechanism for accepting overwriting, it would have to come through some environment variable I think cause this is coming from an event emitted.

We happy with overwriting by default?

Reply:
I'd likely do the opposite in headless mode and default to not overwriting, just due to safety concerns. But I'm not against this proposal either in headless mode for the time being. (As long as we are entirely sure that the folder and files we're touching are REALLY only xgen files and not by change accidentally the root path of a project or whatever, hehe) However, I'd use `log.warning` and also explain that we'll auto-overwrite it due to being in headless/batch mode so it's super clear what's going on. I had even considered that IF it overwrites that maybe it should log which files it's overwriting so the logs make clear that this process touched those files.

Reply:
Yeah, think we should maybe opt for caution here and raise an error explaining which files are in conflict. @iLLiCiTiT are you good with this?

Reply:
It is just quick mock-up, I don't know the consequences. Question is if this can be somehow triggered in headless mode? It is just event callback and the even is triggered only by workflies tool. Maybe raising an exception is right way in this case?

Reply:
> Good point! This only gets triggered from workfiles tool which is always a UI. @BigRoy what do you think of not accounting for headless and just prompt the user, aka discard this suggestion?

Doesn't that just point to a bug that it does not work without workfiles UI? It should, right?
You should instead be using event `before.save` right?

Reply:
> Doesn't that just point to a bug that it does not work without workfiles UI? It should, right?
You should instead be using event before.save right?

We'll need to know where the user is intending on saving the workfile to, because we need to move xgen files and modify attribute before saving.
I also think that modifying the behaviour of Maya native saving with xgen files management will limit the users too much, ei. they cant escape the xgen files management while Maya in launched through the pipeline.